Add a flag to toggle class-templating-syntax

Bpolitycki opened the pull request:
This commit / PR adds a simple setting, which can be toggled by the configuration as part of the templating-model, to control if the old class based templating syntax (class='foo:bar') is interpreted as a call to a xquery function. This will allow to usage of colons inside classes (e.g. to use TailwindCSS)

Bpolitycki committed:
Added a flag to toggle class-templating-syntax

This commit adds a simple setting, which can be toggled by the configuration as part of the templating-model, to control if the old class based templating syntax (class='foo:bar') is interpreted as a call to a xquery function. This will allow to usage of colons inside classes (e.g. to use TailwindCSS)

duncdrum commented:
The feature seems useful to me. Could you please add a test case and address the failing tests. We also need to document the feature in the templating article in documentation repo.

Bpolitycki commented:
> The feature seems useful to me. Could you please add a test case and address the failing tests. We also need to document the feature in the templating article in documentation repo.

Great. Added tests and fixed the problem with the tests (I don't why, but I deleted the casting to booleans before).

So the place for the documentation would be over here: [Templating Docs](https://github.com/eXist-db/documentation/blob/06ef4fb174cae06ca49b97555842792a0d5483da/src/main/xar-resources/data/templating/templating.xml)?

duncdrum commented:
@Bpolitycki yes that is the spot. I ll try to take a look at the CI config later today. Your changes don't seem to have caused all these failures.

Bpolitycki commented:
> @Bpolitycki yes that is the spot. I ll try to take a look at the CI config later today. Your changes don't seem to have caused all these failures.

Perfect – thanks. When running the tests locally, all tests are passing.

Bpolitycki commented:
I think there is a mistake in my approach (which was not covered by the tests). 

https://github.com/Bpolitycki/exist-db-templating/blob/3fae54424b43676a1dbff6a5c04ba4cc15db6862/content/templates.xqm#L152C25-L152C25 

Will stop the execution of templates on childs.

Bpolitycki committed:
Fix recursive template execution for child nodes

This commit introduces a helper function (`templates:process-children`), which will ensure that all templates are called recursive. This is a fixes the bug introduces in the previous feature commit.

line-o commented on package.json:
```diff
@@ -1,56 +1,56 @@
-		"glob-stream": "$glob-stream"
-	}
+  "name": "templating",
+  "version": "1.2.0",
```
Please do not update the version in a PR.
This should be done upon release only.

Bpolitycki replied:
Reverted and fixes. Changes made to `package.json` were removed from the commit.

line-o commented on content/templates.xqm:
```diff
@@ -9,8 +10,7 @@ xquery version "3.1";
 module namespace templates="http://exist-db.org/xquery/html-templating";
-
-import module namespace inspect="http://exist-db.org/xquery/inspection";
+import module namespace inspect="http://exist-db.org/xquery/inspection" at "java:org.exist.xquery.functions.inspect.InspectionModule";
```
The import did work without `at` before. Was it necessary to add it?

Bpolitycki replied:
Fixed. Removed from commit.

line-o commented on test/mocha/rest_spec.js:
```diff
@@ -11,17 +11,17 @@ const { origin } = new URL(serverInfo.server);
 
 const axiosInstance = axios.create({
-  baseURL: app
+  baseURL: app,
```
The comma is not necessary here.

line-o replied:
It should not be part of this PR. See my comment on formatting above.

line-o commented on content/templates.xqm:
```diff
@@ -136,21 +138,26 @@ declare function templates:process($nodes as node()*, $model as map(*)) {
                 $node
 };
 
+declare %private function templates:process-childs($node as node(), $model as map(*)) {
```
Please rename new function to `templates:process-children`.

Bpolitycki replied:
Done.

line-o commented on content/templates.xqm:
```diff
@@ -136,21 +138,26 @@ declare function templates:process($nodes as node()*, $model as map(*)) {
                     if ($dataAttr) then
                         templates:call($dataAttr, $node, $model)
-                    else
+                    else if (($model($templates:CONFIGURATION)($templates:CONFIG_USE_CLASS_SYNTAX), $templates:SEARCH_IN_CLASS)[1]) then
```
`$templates:CONFIG_USE_CLASS_SYNTAX` should be set to `true()` in default configuration.
This will make the condition here much simpler as we can be sure this key is always set.

https://github.com/eXist-db/templating/blob/c0f6e4b45abfba8e79cc69ef5157515d0a847efa/content/templates.xqm#L93

line-o replied:
```
else if ($model?($templates:CONFIGURATION)?($templates:CONFIG_USE_CLASS_SYNTAX)) then

Bpolitycki replied:
Implemented in the suggested way. See https://github.com/eXist-db/templating/pull/25/commits/0a9b651f4d7a15ee32584f7acfe3b8499afc2e1e

line-o commented on test/app/modules/view.xql:
```diff
@@ -111,10 +111,17 @@ declare function test:custom-model($node as node(), $model as map(*)) {
     $templates:CONFIG_APP_ROOT : $test:app-root,
-    $templates:CONFIG_STOP_ON_ERROR: true()
+    $templates:CONFIG_STOP_ON_ERROR: true(),
+    $templates:CONFIG_USE_CLASS_SYNTAX: xs:boolean(request:get-parameter('classLookup', $templates:SEARCH_IN_CLASS))
```
I would prefer to have different configurations to be checked as part of the unit tests. (see
https://github.com/eXist-db/templating/blob/c0f6e4b45abfba8e79cc69ef5157515d0a847efa/test/xqs/test-suite.xqm#L61 and following)

line-o reviewed (changes requested):
Thanks for your contribution. I like the feature.
Here are some things I would like to see:
- I would prefer if we test different configurations as part of the [unit tests](https://github.com/eXist-db/templating/blob/c0f6e4b45abfba8e79cc69ef5157515d0a847efa/test/xqs/test-suite.xqm)
- Please separate commits into changes, fixes and formatting where possible otherwise it is really hard to see what actually changed.
- With which tool did you reformat the JS code and with what settings? We should be on the same page as to not constantly reformat those sources. Maybe best is to leave it out of this PR and first find common ground and set that as a separate build step.

line-o commented:
@Bpolitycki If you need help or have further question, don't hesitate to ask here or on our community slack channel.

Bpolitycki commented:
> @Bpolitycki If you need help or have further question, don't hesitate to ask here or on our community slack channel.

@line-o  Thanks for you feedback – I will have a look at the comments and try to fix the marked issued.

Bpolitycki committed:
Added `$templates:CONFIG_USE_CLASS_SYNTAX` to the default templating-configuration

Bpolitycki commented:
@line-o – I'm really sorry, but it took quite a while to come back on this PR again. 

I've worked on some of the requested changes – but a few are left open.

1. I think the changes to the JS-formatting was done by VSCode (Prettier) – which formatter do you prefer? Let me know and I can fix the formatting. 
2. The test setup is a bit confusing for me. What's the purpose of splitting tests between JS and XQuery? Maybe you can give me an additional hint on how to implement the requested test scenario. 

Thanks in advance!

line-o commented:
@Bpolitycki Tests are split for two reasons:
- XQuery: unit tests, allow to test different configurations
- JS: integration tests

line-o commented:
I, personally, do prefer [standard](https://standardjs.com). It is used in [node-exist](https://github.com/exist-db/node-exist/) and [xst](https://github.com/exist-db/xst/).

line-o commented:
I will try to push a version without any formatting changes so that it is easier for us to develop on top of this change.

line-o commented:
superseded by #29

joewiz commented:
@Bpolitycki Thank you for your contribution!
